jumanji/environments/routing/macvrp/utils.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple
+
+import chex
+import jax
+
+from jumanji.environments.routing.macvrp.constants import DEPOT_IDX
+
+
+def compute_time_penalties(
+    local_times: chex.Array,
+    window_start: chex.Array,
+    window_end: chex.Array,
+    early_coefs: chex.Array,
+    late_coefs: chex.Array,
+) -> chex.Array:
+    """Calculate the time penalties for this step."""
+
+    # Calculate the time penalties for this step.
+    # The time penalties are calculated as the sum of the early and late penalties.
+    # The early penalty is calculated as the difference between the local time and the window
+    # start time multiplied by the early coefficient. The late penalty is calculated as the
+    # difference between the window end time and the local time multiplied by the late coefficient.
+    # The early and late penalties are only calculated if the local time is outside the window.
+    # If the local time is inside the window, the early and late penalties are zero.
```

Comment:
Could we include these comments inside the docstring?

Reply:
Updated.